storage/remote/queue_manager.go

```diff
@@ -1028,7 +1032,89 @@ func (s *shards) enqueue(ref chunks.HeadSeriesRef, data interface{}) bool {
 	}
 }
 
-func (s *shards) runShard(ctx context.Context, shardID int, queue chan interface{}) {
+type queue struct {
+	batch      []interface{}
+	batchQueue chan []interface{}
+
+	// Since we know there is a limited number of batches out, using a stack is
+	// easy and persistent across GC compared to a sync.Pool
+	batchPool [][]interface{}
+	poolMux   sync.Mutex
+}
+
+func newQueue(batchSize, capacity int) *queue {
+	batches := capacity / batchSize
+	return &queue{
+		batch:      make([]interface{}, 0, batchSize),
+		batchQueue: make(chan []interface{}, batches),
+		batchPool:  make([][]interface{}, 0, batches),
+	}
+}
+
+func (q *queue) Append(datum interface{}, stop <-chan struct{}) bool {
+	q.batch = append(q.batch, datum)
+	if len(q.batch) == cap(q.batch) {
+		select {
+		case q.batchQueue <- q.batch:
+			q.batch = q.newBatch(cap(q.batch))
+			return true
+		case <-stop:
+			// Remove the sample we just appended. It will get retried.
+			q.batch = q.batch[:len(q.batch)-1]
+			return false
+		}
+	}
+	return true
+}
+
+func (q *queue) Chan() <-chan []interface{} {
+	return q.batchQueue
+}
+
+// Batch returns the current batch and allocates a new batch. Must not be
+// called concurrently with Append.
+func (q *queue) Batch() []interface{} {
+	batch := q.batch
+	q.batch = q.newBatch(cap(batch))
+	return batch
+}
+
+func (q *queue) Return(batch []interface{}) {
```

Comment:
this function name isn't clear to me, maybe it should be `Clear` or something similar?

Reply:
Or just a comment `returns the batch buffer to the internal pool`

Reply:
Sounds good, I added the comment for now, but would be happy to rename the function as well. We could do something like ReturnBatch to make it more specific? Or PoolBatch?

Reply:
`ReturnForReuse`?